src/telemetry.ts

```diff
@@ -36,4 +49,80 @@ export class NullTelemetryReporter implements ITelemetryReporter {
 
 }
 
+export const DefaultTelemetryInterval = 10000;
+
+export class BatchTelemetryReporter implements ITelemetryReporter {
+    private _eventBuckets: {[eventName: string]: any};
+    private _timer: NodeJS.Timer;
+
+    public constructor(private _telemetryReporter: TelemetryReporter, private _cadence: number = DefaultTelemetryInterval) {
+        this.reset();
+        this.setup();
+    }
+
+    public reportEvent(name: string, data?: any): void {
+        if (!this._eventBuckets[name]) {
+            this._eventBuckets[name] = [];
+        }
+
+        this._eventBuckets[name].push(data);
+    }
+
+    public setupEventHandler(_sendEvent: (event: DebugProtocol.Event) => void): void {
+        // no-op
+    }
+
+    public finalize(): void {
+        this.send();
+        clearInterval(this._timer);
+    }
+
+    private setup(): void {
+        this._timer = setInterval(() => this.send(), this._cadence);
+    }
+
+    private reset(): void {
+        this._eventBuckets = {};
+    }
+
+    private send(): void {
```

Comment:
I'm finding hard to follow what this method does. I think it'd be much easier to understand if we split this into smaller private methods with meaningful self-documenting names. What do you think?

Reply:
It's been divided into several parts.